File: backend/game_db/Tower/frontend/src/map/engagement.ts
import type { Army, Engagement, HexCoord, Structure, TowerMap } from "../types";
import { coordKey, tileAt } from "./routing";

// Frontend mirror of engine/engagement.py so the sandbox can surface strategic
// combat triggers (army collision / structure assault / control contest) without
// a round trip to the Python engine.  The Python engine remains the source of
// truth; this only previews the same conditions on the generated layout.

function opponent(side: "human" | "monster"): "human" | "monster" {
  return side === "human" ? "monster" : "human";
}

function hexRange(center: HexCoord, radius: number): HexCoord[] {
  const out: HexCoord[] = [];
  for (let dq = -radius; dq <= radius; dq += 1) {
    const lo = Math.max(-radius, -dq - radius);
    const hi = Math.min(radius, -dq + radius);
    for (let dr = lo; dr <= hi; dr += 1) {
      out.push({ q: center.q + dq, r: center.r + dr });
    }
  }
  return out;
}

function controlledTiles(structure: Structure): Set<string> {
  const out = new Set<string>();
  for (const coord of structure.footprint) {
    for (const candidate of hexRange(coord, structure.controlRadius)) {
      out.add(coordKey(candidate));
    }
  }
  return out;
}

function structureAt(map: TowerMap, coord: HexCoord): Structure | undefined {
  const tile = tileAt(map, coord);
  if (!tile || !tile.structureId) return undefined;
  return map.structures.find((item) => item.id === tile.structureId);
}

export function detectEngagements(map: TowerMap, month = 1): Engagement[] {
  const engagements: Engagement[] = [];
  const seen = new Set<string>();

  const push = (
    location: HexCoord,
    attackers: Army[],
    defenders: Army[],
    attackerSide: "human" | "monster" | null,
    defenderSide: "human" | "monster" | null,
    structureId: string | null,
    reason: Engagement["reason"]
  ): void => {
    if (!attackers.length) return;
    if (!defenders.length && !structureId) return;
    const attackerIds = attackers.map((army) => army.id).sort();
    const defenderIds = [...defenders.map((army) => army.id), ...(structureId ? [structureId] : [])].sort();
    const dedupe = `${coordKey(location)}|${attackerIds.join(",")}|${defenderIds.join(",")}`;
    if (seen.has(dedupe)) return;
    seen.add(dedupe);
    engagements.push({
      id: `${coordKey(location)}:${reason}:${engagements.length}`,
      location,
      month,
      attackers: attackerIds,
      defenders: defenders.map((army) => army.id).sort(),
      attackerSide,
      defenderSide,
      structureId,
      reason
    });
  };

  // 1) enemy armies sharing a tile
  const byTile = new Map<string, Army[]>();
  for (const army of map.armies) {
    const tileKey = coordKey(army.position);
    const list = byTile.get(tileKey) ?? [];
    list.push(army);
    byTile.set(tileKey, list);
  }
  for (const armies of byTile.values()) {
    if (new Set(armies.map((army) => army.owner)).size <= 1) continue;
    const attackerSide = [...armies].sort((a, b) => a.owner.localeCompare(b.owner))[0].owner;
    const defenderSide = opponent(attackerSide);
    push(
      armies[0].position,
      armies.filter((army) => army.owner === attackerSide),
      armies.filter((army) => army.owner === defenderSide),
      attackerSide,
      defenderSide,
      null,
      "army_collision"
    );
  }

  // 2) army standing on an enemy-owned structure
  for (const army of map.armies) {
    const structure = structureAt(map, army.position);
    if (structure && structure.owner && structure.owner !== army.owner) {
      push(army.position, [army], [], army.owner, structure.owner, structure.id, "structure_assault");
    }
  }

  // 3) army inside an enemy structure's control radius
  const controlled = map.structures
    .filter((structure) => structure.owner)
    .map((structure) => ({ structure, coords: controlledTiles(structure) }));
  for (const army of map.armies) {
    for (const { structure, coords } of controlled) {
      if (structure.owner !== army.owner && coords.has(coordKey(army.position))) {
        push(army.position, [army], [], army.owner, structure.owner, structure.id, "control_contest");
      }
    }
  }

  return engagements;
}

File: backend/game_db/Tower/engine/engagement.py
"""M2 strategic engagement detection."""

from __future__ import annotations

from collections import defaultdict

from schema.models import Army, ArmyMove, DeployIntent, Engagement, GameState, HexCoord, Side, Structure, controlled_tiles, derive_seed


def _opponent(side: Side) -> Side:
    return Side.MONSTER if side == Side.HUMAN else Side.HUMAN


def _engagement_id(state: GameState, coord: HexCoord, index: int) -> str:
    return derive_seed(state.master_seed, "engagement", state.turn.month, coord.key(), index)


def _split_armies(armies: list[Army], attacker_side: Side | None = None) -> tuple[list[Army], list[Army], Side | None, Side | None]:
    if attacker_side is None:
        sides = sorted({army.owner for army in armies}, key=lambda item: item.value)
        attacker_side = sides[0] if sides else None
    defender_side = _opponent(attacker_side) if attacker_side is not None else None
    attackers = [army for army in armies if army.owner == attacker_side]
    defenders = [army for army in armies if army.owner == defender_side]
    return attackers, defenders, attacker_side, defender_side


def _structure_at(state: GameState, coord: HexCoord) -> Structure | None:
    tile = state.tile_at(coord)
    if tile is None or tile.structure_id is None:
        return None
    return state.structures.get(tile.structure_id)


def find_engagements(state: GameState, moves: list[ArmyMove] | None = None) -> list[Engagement]:
    """Find combat triggers after movement has been applied.

    M2 only creates strategic triggers.  M4 will expand each engagement into a
    tactical battle map and deterministic battle replay.
    """
    engagements: list[Engagement] = []
    seen: set[tuple[str, str, str]] = set()

    def add(
        coord: HexCoord,
        armies: list[Army],
        reason: Engagement["reason"],
        attacker_side: Side | None = None,
        structure: Structure | None = None,
    ) -> None:
        attackers, defenders, resolved_attacker, resolved_defender = _split_armies(armies, attacker_side)
        if structure is not None and structure.owner is not None:
            resolved_defender = structure.owner
            resolved_attacker = _opponent(structure.owner) if resolved_attacker is None else resolved_attacker
            attackers = [army for army in armies if army.owner == resolved_attacker]
            defenders = [army for army in armies if army.owner == resolved_defender]
        if not attackers:
            return
        if not defenders and structure is None:
            return
        key = (
            coord.key(),
            ",".join(sorted(army.id for army in attackers)),
            ",".join(sorted([army.id for army in defenders] + ([structure.id] if structure else []))),
        )
        if key in seen:
            return
        seen.add(key)
        engagements.append(
            Engagement(
                id=_engagement_id(state, coord, len(engagements)),
                location=coord,
                month=state.turn.month,
                attackers=sorted(army.id for army in attackers),
                defenders=sorted(army.id for army in defenders),
                attacker_side=resolved_attacker,
                defender_side=resolved_defender,
                structure_id=structure.id if structure is not None else None,
                reason=reason,
            )
        )

    by_coord: dict[str, list[Army]] = defaultdict(list)
    for army in state.armies.values():
        by_coord[army.position.key()].append(army)
    for armies in by_coord.values():
        if len({army.owner for army in armies}) > 1:
            add(armies[0].position, armies, "army_collision")

    for army in state.armies.values():
        structure = _structure_at(state, army.position)
        if structure is not None and structure.owner is not None and structure.owner != army.owner:
            add(army.position, [army], "structure_assault", attacker_side=army.owner, structure=structure)

    controlled: list[tuple[Structure, set[str]]] = [
        (structure, {coord.key() for coord in controlled_tiles(structure)})
        for structure in state.structures.values()
        if structure.owner is not None
    ]
    for army in state.armies.values():
        for structure, coords in controlled:
            if structure.owner != army.owner and army.position.key() in coords:
                add(army.position, [army], "control_contest", attacker_side=army.owner, structure=structure)

    for move in moves or []:
        if move.intent != DeployIntent.ATTACK:
            continue
        army = state.armies.get(move.army_id)
        if army is None:
            continue
        local_armies = [item for item in state.armies.values() if item.position == army.position]
        structure = _structure_at(state, army.position)
        if any(item.owner != army.owner for item in local_armies) or (
            structure is not None and structure.owner is not None and structure.owner != army.owner
        ):
            add(army.position, local_armies, "attack_order", attacker_side=army.owner, structure=structure)

    return engagements
